flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2113,6 +2122,63 @@ IntrinsicLibrary::genRepeat(mlir::Type resultType,
       });
 }
 
+// RESHAPE
+fir::ExtendedValue
+IntrinsicLibrary::genReshape(mlir::Type resultType,
+                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 4);
+
+  // Handle source argument
+  auto source = builder.createBox(loc, args[0]);
+
+  // Handle shape argument
+  auto shape = builder.createBox(loc, args[1]);
+  fir::BoxValue shapeTmp = shape;
+  auto shapeRank = shapeTmp.rank();
+  assert(shapeRank == 1);
+  auto shapeTy = shape.getType();
+  auto shapeArrTy = fir::dyn_cast_ptrOrBoxEleTy(shapeTy);
+  auto shapeSeqTy = shapeArrTy.cast<fir::SequenceType>();
+  auto shapeEleTy = shapeSeqTy.getEleTy();
+  assert(fir::isa_integer(shapeEleTy));
+  auto resultRank = shapeSeqTy.getShape();
+
+  // Handle optional pad argument
+  auto pad = isAbsent(args[2])
+                 ? builder.create<fir::AbsentOp>(
+                       loc, fir::BoxType::get(builder.getI1Type()))
+                 : builder.createBox(loc, args[2]);
+
+  // Handle optional order argument
+  auto order = isAbsent(args[3])
+                   ? builder.create<fir::AbsentOp>(
+                         loc, fir::BoxType::get(builder.getI1Type()))
+                   : builder.createBox(loc, args[3]);
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+  auto type = builder.getVarLenSeqTy(resultType, resultRank[0]);
```

Comment:
Given the `SIZE(shape)` is a constant constraint, it should be the case that the extent is constant in shape type, but lowering cannot completely guarantee that it won't lose track of constant shape information. I think right now you are not seeing troubles because `asAddr` tends to better preserve the constant extents on array-sections (because the temp is created using the front-end `GetShape(expr)` that is guaranteed to return a constant shape), but I am afraid it is a bit fragile.

You should at least add a fir::emitFatalError if resultRank is "unknown extent" (see SequenceType definition in `include/flang/Optimizer/Dialect/FIRTypes.td`) for now. The ideal would be to find a safe way to propagate the result rank from GetShape(expr) here for the result. In the meantime, you should stick to `asAddr` for the `shape` argument (it will always be a small array anyway (extent <= 15), so a copy is not too bad for potential array-sections here for now).

Reply:
The ORDER argument has the same shape as the SHAPE argument. So, I assume ORDER should remain as an "asAddr" argument too?

Reply:
No, that we do not care because we do not need to get the size out-of it. The point of having `SHAPE` lowered with `asAddr` is more of a workaround to be sure to be able to compute the rank of the result.

We do not really care about being able to verify here that SHAPE and ORDER have the same shape. If ORDER shape is constant, the front-end should complain in case of shape mismatch, and in case it's not, the runtime is enforcing this at runtime.

Reply:
> Given the `SIZE(shape)` is a constant constraint, it should be the case that the extent is constant in shape type, but lowering cannot completely guarantee that it won't lose track of constant shape information. I think right now you are not seeing troubles because `asAddr` tends to better preserve the constant extents on array-sections (because the temp is created using the front-end `GetShape(expr)` that is guaranteed to return a constant shape), but I am afraid it is a bit fragile.

If lowering is "forgetting" constant shape information, lowering has bugs that need to be fixed. Furthermore, fixing such bugs should take priority over designing workarounds so the bugs can linger around longer and not be fixed.

Reply:
@mleair, this discussion impacts reshape lowering somehow, but I do not think you need to wait to do anything about it in your PR yet.

> If lowering is "forgetting" constant shape information, lowering has bugs that need to be fixed.

I am not sure how much I agree this is a bug. To me, the facts that fir array types of expression/sub-expression with constant shape according to Fortran rules actually are fir.array types with constant shape is a very important feature for the optimizer that we want to get right 99% of the time, which would not sufficient to rely on it for correctness purpose.

If we want to get it right 100% of the case, we need to do something to enforce this by design in array expression lowering. Right, now the result type of a temporary array created with `createSomeArrayTempValue` are by design guarantee to get constant shape when possible because `evaluate::GetShape()` result is used if it is constant to allocate the result, which right now is the more robust tool to compute constant shapes. That leaves a lot of contexts where array expression lowering is currently not enforcing this feature, the main one being `createSomeArrayBox` that is not getting constant shaped fir.box out of constant shaped array-sections. We would also need to build constant shaped types when possible for all the temp result fir.box that are passed to the runtime for allocation. 

So I am not against it, I just want to point that it is something that needs to be well thought and designed in lowering rather than a few bugs to be fixed.